rabbit-openqa.py

```diff
@@ -0,0 +1,209 @@
+            if entry.get('name').endswith('iso'):
+                ret.append(self.map_iso(name, entry.get('name')))
+        return ret
+
+    def initial_staging_state(self, name):
+        ret = {'isos': self.gather_isos(name, 'images')}
+        # missing API for initial repo id
+        return ret
+
+    def update_staging_buildid(self, project, repository, buildid):
+        self.staging_projects[project]['id'] = buildid
+        self.staging_projects[project]['isos'] = self.gather_isos(project, repository)
+        print('UPDATE', project, self.staging_projects[project])
+
+    def check_published_repo(self, project, repository, buildid):
+        if repository != 'images':
+            return
+        for p in self.staging_projects:
+            if project == p:
+                self.update_staging_buildid(project, repository, buildid)
+
+    def matching_project(self, iso):
+        for p in self.staging_projects:
+            if iso in self.staging_projects[p]['isos']:
+                return p
+
+    def map_openqa_result(self, result):
+        if result in ['passed', 'softfailed']:
+            return 'success'
+        return 'failure'
+
+    def openqa_done(self, iso, test, machine, id, result):
+        print('openqa_done', iso, test, machine, id, result)
+        staging = self.matching_project(iso)
+        if not staging:
+            return
+        buildid = self.staging_projects[staging].get('id')
+        if not buildid:
+            print("I don't know the build id of " + staging)
+            return
+        xml = self.openqa_check_xml(id,
+                                    self.map_openqa_result(result),
+                                    test + '@' + machine)
+        url = self.api.makeurl(['status_reports', 'published', staging, 'images', 'reports', buildid])
+        http_POST(url, data=xml)
+    
+    def openqa_create(self, iso, test, machine, id):
+        print('openqa_create', iso, test, machine, id)
+
+    def openqa_check_xml(self, id, state, name):
+        check = ET.Element('check')
+        se = ET.SubElement(check, 'url')
+        se.text = "https://openqa.suse.de/tests/{}".format(id)
+        se = ET.SubElement(check, 'state')
+        se.text = state
+        se = ET.SubElement(check, 'name')
+        se.text = name
+        return ET.tostring(check)
+
+class Listener(object):
+    def __init__(self, amqp_prefix, amqp_url):
+        self.projects = []
+        self.amqp_prefix = amqp_prefix
+        self.amqp_url = amqp_url
+        connection = pika.BlockingConnection(pika.URLParameters(amqp_url))
+        self.channel = connection.channel()
+
+        self.channel.exchange_declare(exchange='pubsub', exchange_type='topic', passive=True, durable=True)
+
+        result = self.channel.queue_declare(exclusive=True)
+        queue_name = result.method.queue
+
+        self.channel.queue_bind(exchange='pubsub',
+                                queue=queue_name,routing_key='#')
```

Comment:
Binding to events of interest (listed in `on_message`) should seem preferable.

Reply:
Yes. As the title says: WIP

Reply:
And this queue bind would happen server side (see above) and we would consume everything of *that* queue

Reply:
Now that I implemented a cold start, I decided against using a custom queue and just bind the 4 routing keys on pubsub.